access-control-service-sql-impl/src/main/java/org/eclipse/tractusx/semantics/accesscontrol/sql/service/SqlBackedAccessControlRuleService.java

```diff
@@ -65,9 +67,11 @@ public List<String> filterValidSpecificAssetIdsForLookup(
                      .filter( accessControlRule -> aShellContext.specificAssetIds().containsAll( accessControlRule.getMandatorySpecificAssetIds() ) )
                      .flatMap( accessControlRule -> accessControlRule.getVisibleSpecificAssetIdNames().stream() )
                      .collect( Collectors.toSet() );
+               Map<String, Set<String>> duplicatedSpecificAssetIds = getDuplicateSpecificAssetIds( aShellContext );
                return aShellContext.specificAssetIds().stream()
-                     .filter( id -> visibleSpecificAssetIdNames.contains( id.name() ) )
-                     .collect( Collectors.toSet() ).containsAll( userQuery );
+                     .filter( id -> isSpecificAssetIdVisible( id, visibleSpecificAssetIdNames, duplicatedSpecificAssetIds ) )
```

Comment:
Just making a note about what we discussed on Teams. The mandatory Ids of the rules are lost this way, so it can skew the comparison. We might need to use a similar approach as below, building a multi-map and adding not only the visible ids, but the mandatory as well before evaluating the visibility of the shell.